doc/man/man1/virgilconfirm.1
.TH "VIRGIL" "1" "August 01, 2015" "Virgil Security CLI (0.3.0)" "BSD General Commands Manual"
.SH NAME
.PP
\f[B]confirm\f[] \-\- send confirmation code to the Virgil Public Keys
service.
.SH SYNOPSIS
.PP
\f[B]virgil confirm\f[] \-\-user\-id \f[I]arg\f[] \-\-code \f[I]arg\f[]
.SH DESCRIPTION
.PP
Send confirmation code to the Virgil Public Keys service.
Confirmation code is sent to the user\[aq]s identifier (email, phone,
etc) after public key registration.
.SH OPTIONS
.TP
.B \-i \f[I]arg\f[], \-\-user\-id \f[I]arg\f[]
(required) User\[aq]s identifer.
Format:
.RS
.IP
.nf
\f[C]
[email|phone|domain]:<value>
\f[]
.fi
.PP
where:
.IP \[bu] 2
if \f[C]email\f[], then \f[I]value\f[] \- user\[aq]s email;
.IP \[bu] 2
if \f[C]phone\f[], then \f[I]value\f[] \- user\[aq]s phone;
.IP \[bu] 2
if \f[C]domain\f[], then \f[I]value\f[] \- user\[aq]s domain.
.RE
.TP
.B \-c \f[I]arg\f[], \-\-code \f[I]arg\f[]
(required) Confirmation code.
.RS
.RE
.SH SEE ALSO
.PP
\f[C]virgil(1)\f[], \f[C]virgilkeyreg(1)\f[]
.SH AUTHORS
(c) Virgil Security Inc.


ChangeLog
virgil CLI ChangeLog (Sorted per date)

= Version 0.3.0 released 2015-08-13

Improvements:
  * Add Markdown documentation.

Changes:
  * Replace certificate term with Virgil Public Key term.
  * Remove options to specify sign's output format.
  * Use C++11 standard.

Features:
  * Add command "keyreg" - register user's public key on the Virgil Public Keys service.
  * Add command "keyget" - get user's public key from the Virgil Public Keys service.
  * Add command "confirm" - send confirmation code to the Virgil Public Keys service.
  * Add command "reconfirm" - resend confirmation code to the user for given user's identifier.

= Version 0.2.0 released 2015-05-15

Bugfix:
  * Use standard constructor of class 'ifstream'.
  * Change parameter 'p' long name from 'arg' to 'pwd'.
  * Add required include statement <cstdlib>.

Improvements:
  * Fix misspelling in the help text.
  * Add "happy path" test.
  * Modify 'encrypt' and 'verify' commands to work with Virgil PKI service thru user's identifier.

Changes:
  * Change parameter 'asn1' to 'der'.
  * Change parameters format for 'encrypt' command.

Features:
  * Add 'verify' command.
  * Add 'adduser' command

= Version 0.1.0 released 2015-05-04

Features:
  * Add command "keygen" - generate private key with given parameters.
  * Add command "key2pub" - extract public key from the private key.
  * Add command "pub2cert" - create certificate from the public key and identifiers.
  * Add command "certinfo" - output detailed information about given certificate.
  * Add command "encrypt" - encrypt data for given recipients which can be defined by certificates and by passwords.
  * Add command "decrypt" - decrypt data for given recipient which can be defined by certificate or by password.
  * Add command "sign" - sign data with private key.
  * Add command "verify" - verify data with certificate.


src/commands/virgilconfirm.cxx
/**
 * Copyright (C) 2015 Virgil Security Inc.
 *
 * Lead Maintainer: Virgil Security Inc. <support@virgilsecurity.com>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 *
 *     (1) Redistributions of source code must retain the above copyright
 *     notice, this list of conditions and the following disclaimer.
 *
 *     (2) Redistributions in binary form must reproduce the above copyright
 *     notice, this list of conditions and the following disclaimer in
 *     the documentation and/or other materials provided with the
 *     distribution.
 *
 *     (3) Neither the name of the copyright holder nor the names of its
 *     contributors may be used to endorse or promote products derived from
 *     this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <iostream>
#include <fstream>
#include <stdexcept>
#include <string>

#include <tclap/CmdLine.h>

#include <virgil/sdk/keys/http/Connection.h>
#include <virgil/sdk/keys/model/UserData.h>
#include <virgil/sdk/keys/client/KeysClient.h>

#include <cli/version.h>
#include <cli/config.h>
#include <cli/util.h>
#include <cli/pair.h>
#include <cli/guid.h>

using virgil::sdk::keys::http::Connection;
using virgil::sdk::keys::model::UserData;
using virgil::sdk::keys::client::KeysClient;


#ifdef SPLIT_CLI
    #define MAIN main
#else
    #define MAIN confirm_main
#endif


int MAIN(int argc, char **argv) {
    try {
        // Parse arguments.
        TCLAP::CmdLine cmd("Send confirmation code to the Virgil Public Keys service. "
                "Confirmation code is sent after user's public key registration.", ' ',
                virgil::cli_version());

        TCLAP::ValueArg<std::string> userIdArg("i", "user-id",
                "User's identifer.\n"
                "Format: [email|phone|domain]:<value>\n"
                "where:\n"
                "\t* if email, then <value> - user's email;\n"
                "\t* if phone, then <value> - user's phone;\n"
                "\t* if domain, then <value> - user's domain.\n",
                true, "", "arg");

        TCLAP::ValueArg<std::string> confirmationCodeArg("c", "code", "Confirmation code.",
                true, "", "arg");

        cmd.add(confirmationCodeArg);
        cmd.add(userIdArg);

        cmd.parse(argc, argv);

        // Parse User Identifier
        const std::pair<std::string, std::string> userIdPair = virgil::cli::parse_pair(userIdArg.getValue());
        const std::string userIdType = userIdPair.first;
        const std::string userId = userIdPair.second;

        // Find User Data
        KeysClient keysClient(std::make_shared<Connection>(VIRGIL_APP_TOKEN));
        auto foundUserDatas = keysClient.userData().search(userId, userIdType);

        // Confirm User Data
        if (foundUserDatas.size() > 0) {
            UserData userData = foundUserDatas.front();
            keysClient.userData().confirm(userData.userDataId(), confirmationCodeArg.getValue(), virgil::cli::guid());
        } else {
            throw std::runtime_error("User with id: " + userIdArg.getValue() + " not found.");
        }
    } catch (TCLAP::ArgException& exception) {
        std::cerr << "Error: " << exception.error() << " for arg " << exception.argId() << std::endl;
        return EXIT_FAILURE;
    } catch (std::exception& exception) {
        std::cerr << "Error: " << exception.what() << std::endl;
        return EXIT_FAILURE;
    }

    return EXIT_SUCCESS;
}
